pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,139 +18,254 @@
+        if (!elasticsearchClient.isFailed()) {
+            try {
+                Pair<String, String> idAndDoc = extractIdAndDocument(record);
+                if (idAndDoc.getRight() == null) {
+                    switch (elasticSearchConfig.getNullValueAction()) {
+                        case DELETE:
+                            if (elasticSearchConfig.isBulkEnabled()) {
+                                elasticsearchClient.bulkDelete(record, idAndDoc.getLeft());
+                            } else {
+                                elasticsearchClient.deleteDocument(record, idAndDoc.getLeft());
+                            }
+                            break;
+                        case IGNORE:
+                            break;
+                        case FAIL:
+                            elasticsearchClient.failed(new PulsarClientException.InvalidMessageException("Unexpected null message value"));
+                            throw elasticsearchClient.irrecoverableError.get();
+                    }
+                } else {
+                    if (elasticSearchConfig.isBulkEnabled()) {
+                        elasticsearchClient.bulkIndex(record, idAndDoc);
+                    } else {
+                        elasticsearchClient.indexDocument(record, idAndDoc);
+                    }
+                }
+            } catch (Exception e) {
+                log.error("write error:", e);
+                throw e;
             }
-        } catch (final IOException ex) {
-            record.fail();
         }
     }
 
-    public KeyValue<String, byte[]> extractKeyValue(Record<byte[]> record) {
-        String key = record.getKey().orElse("");
-        return new KeyValue<>(key, record.getValue());
+    @VisibleForTesting
+    ElasticsearchClient getElasticsearchClient() {
+        return this.elasticsearchClient;
     }
 
-    private void createIndexIfNeeded() throws IOException {
-        GetIndexRequest request = new GetIndexRequest();
-        request.indices(elasticSearchConfig.getIndexName());
-        boolean exists = getClient().indices().exists(request, RequestOptions.DEFAULT);
+    /**
+     * Extract ES _id and _source using the Schema if available.
+     *
+     * @param record
+     * @return A pair for _id and _source
+     */
+    public Pair<String, String> extractIdAndDocument(Record<GenericObject> record) {
+        Object key = null;
+        Object value = null;
+        Schema<?> keySchema = null;
+        Schema<?> valueSchema = null;
+
+        if (record.getSchema() instanceof KeyValueSchema) {
+            KeyValueSchema<GenericObject,GenericObject> keyValueSchema = (KeyValueSchema) record.getSchema();
+            keySchema = keyValueSchema.getKeySchema();
+            valueSchema = keyValueSchema.getValueSchema();
+            if (KeyValueEncodingType.SEPARATED.equals(keyValueSchema.getKeyValueEncodingType()) && record.getKey().isPresent()) {
```

Comment:
there should not be a distinction between SEPARATED and INLINE

I suggest to use:
`key = ((KeyValue) record.getValue().getNativeObject()).getKey();`

Reply:
Yes, fixed